### ISSUE TITLE
Make CRAB REST stream all logs to container's stdout

### ISSUE THREAD

**Issue by novicecpp** [dmwm/CRABServer#7463]
Subtask of https://github.com/dmwm/CRABServer/issues/7408

We want to stream all crabserver logs to the container's stdout to deprecate logging on cephfs.
**BUT, crabserver runs as a daemon process, not PID 1 or main process 1. It complicates to stream stdout of the daemonizing process to the stdout of PID 1**
("Main process" refers to the child of PID 1 that run forever. If it gets killed container will be terminated.)

To understand the above statement. Let's start with how crabserver container run.

## How does the container run

![process_execution](https://user-images.githubusercontent.com/2346664/202799447-21ce977d-2d72-42d4-a153-5e623bd282bb.png)

Note:
- Entrypoint is how the container runs the first process (PID 1). It overrides `ENTRYPOINT` of Dockerfile by `command` in [crabserver.yaml](https://github.com/dmwm/CMSKubernetes/blob/e3de55fa56b215dba7deb5d1d49c22a00a6a6936/kubernetes/cmsweb/services/crabserver.yaml#L185).
- Edge arrow line: the parent call the child.
- Box with dot line: Process spawn but terminate later.

Here is the output from `ps` command, which correlates to the graph above

```
[_crabserver@crabserver-695f7d8c55-stxdg data]$ ps -aefj --forest
UID          PID    PPID    PGID     SID  C STIME TTY          TIME CMD
_crabse+  301222       0  301222  301222  0 15:48 pts/0    00:00:00 bash
_crabse+  301352  301222  301352  301222  0 15:48 pts/0    00:00:00  \_ ps -aefj --forest
_crabse+       1       0       1       1  0 Nov17 ?        00:00:00 /bin/bash /opt/setup-certs-and-run/setup-certs-and-run.sh
_crabse+      10       1       1       1  0 Nov17 ?        00:00:00 /bin/bash /data/run.sh
root         146      10       1       1  0 Nov17 ?        00:00:00  \_ sudo /usr/sbin/crond -n
root         171     146       1       1  0 Nov17 ?        00:00:00      \_ /usr/sbin/crond -n
_crabse+     100       1       1       1  0 Nov17 ?        00:00:01 rotatelogs /data/srv/logs/crabserver/crabserver-%Y%m%d-crabserver-695f7d8c55-stxdg.log 86400
_crabse+     103       1     101     101  0 Nov17 ?        00:00:00 python3 /data/srv/HG2211b-cb31bdf17981ab54e1d311df7b1f0e57/sw.crab_master/slc7_amd64_gcc630/cms/crabse
_crabse+     105     103     101     101  0 Nov17 ?        00:02:23  \_ python3 /data/srv/HG2211b-cb31bdf17981ab54e1d311df7b1f0e57/sw.crab_master/slc7_amd64_gcc630/cms/cr
_crabse+     142       1       1       1  0 Nov17 ?        00:00:04 /bin/bash /data/cmsweb/bin/process_monitor.sh /data/srv/state/crabserver/crabserver.pid crabserver :18
_crabse+     229     142       1       1  0 Nov17 ?        00:00:36  \_ process_exporter -pid 105 -prefix crabserver -address :18270
_crabse+  301351     142       1       1  0 15:48 ?        00:00:00  \_ sleep 15
```

Things get interesting when `wmc-httpd` script run, which executes main() of [WMCore/REST/Main.py](https://github.com/dmwm/WMCore/blob/c4055fb1e352efd555844f5edcd2184896efd5cb/src/python/WMCore/REST/Main.py#L508).

When execution line go to [start\_daemon()](https://github.com/dmwm/WMCore/blob/c4055fb1e352efd555844f5edcd2184896efd5cb/src/python/WMCore/REST/Main.py#L389), it does:

- Execute `rotatelogs` as a subprocess and change stdout (plus stdin and stderr) of *current process* to stdout of `rotatelogs`.
- double fork itself to establish itself as daemonize process. I call it `crabserver_daemon`.
- Then fork itself again at [line 464](https://github.com/dmwm/WMCore/blob/c4055fb1e352efd555844f5edcd2184896efd5cb/src/python/WMCore/REST/Main.py#L464), which is serving the request. I call the child process `crabserver_serve`.

After `wmc-httpd` executes double fork, the original `wmc-httpd` process get terminated, and [run.sh](https://github.com/dmwm/CMSKubernetes/blob/e3de55fa56b215dba7deb5d1d49c22a00a6a6936/docker/crabserver/run.sh) continue execution and call [monitoring.sh](https://github.com/dmwm/CMSKubernetes/blob/e3de55fa56b215dba7deb5d1d49c22a00a6a6936/docker/crabserver/monitor.sh), then finally call `crond`.

## Problem statement

In the current configuration, we pipe stdout to `rotatelogs` process and let it save to files and rotate the file for us ([manage at line 54](https://github.com/dmwm/deployment/blob/88e57d864473f7557884bdc8da947639c4e2f091/crabserver/manage#L54)).

(Note that cherrypy is configured streaming logs to stdout.)

The question is **"How do we remove pipe stdout to `rotatelogs` and pipe it to stdout of PID 1 instead?"**

### 1. Use stdout inherit from fork()

Basically, we replace [WMCore/REST/Main.py](https://github.com/dmwm/WMCore/blob/c4055fb1e352efd555844f5edcd2184896efd5cb/src/python/WMCore/REST/Main.py#L401-L408) with ` logger = sys.stdout`.

And take advantage of calling child process and `fork()` that child process inherit stdout/stderr from the parent (need citation) to stream logs to the same stdout as PID 1.

The disadvantage of this solution is when we restart crabserver with [/data/start.sh](https://github.com/dmwm/CMSKubernetes/blob/e3de55fa56b215dba7deb5d1d49c22a00a6a6936/docker/crabserver/start.sh), i.e., to run in GH mode, it will use your shell stdout instead.

You still see the logs, but it not recorded anywhere in the k8s server.

### 2. Use named pipe

Named pipe is similar to [pipe](https://man7.org/linux/man-pages/man7/pipe.7.html) in shell (e.g., `cat *.log | grep ERROR`) but create it as a file in filesystem. It is somewhat like a pointer to an actual pipe. It does not do any rw filesystem operation.
To use, you need at least 2 processes, one for reading and another one for writing. That means process A cannot write to named pipe until process B starts reading it.

- In [run.sh before line 67](https://github.com/dmwm/CMSKubernetes/blob/e3de55fa56b215dba7deb5d1d49c22a00a6a6936/docker/crabserver/run.sh#L67),
  - create named pipe using `mkfifo`. i.e., `mkfifo /mypipe`
  - and at the end of [run.sh at line 77](https://github.com/dmwm/CMSKubernetes/blob/e3de55fa56b215dba7deb5d1d49c22a00a6a6936/docker/crabserver/run.sh#L77), instead of running `crond`, we run `cat /mypipe` inside infinte loop.
- In [manage line 54](https://github.com/dmwm/deployment/blob/88e57d864473f7557884bdc8da947639c4e2f091/crabserver/manage#L54), we pass named pipe's path.
- In [WMCore/REST/Main.py at line 406](https://github.com/dmwm/WMCore/blob/c4055fb1e352efd555844f5edcd2184896efd5cb/src/python/WMCore/REST/Main.py#L406), change mode to `a+b`. Otherwise, it will raise the exception "ValueError: can't have unbuffered text I/O".

Please see the files change above in https://github.com/novicecpp/CRABServer/tree/logs_stdout/cicd/logs_stdout

This way, logs that output to stdout/stderr always stream to the container's stdout, through named pipe and `cat` process.

---------

I would like to thanks @mapellidario for brainstorming session. Feel free to have questions or comments.

@belforte I would like to propose solution 2) for streaming log to container's stdout. What do you think?

**Comment by belforte** [comment]
let's have a chat after a bit of time to digest. Maybe there are some typos which confuse me. Better by voice.
My first priority here is to keep things simple.

questions af first sigh:

in 1), can we work around the problem with GH mode by doing `./start.sh -g | tee  /data/srv/logs/crabserver/somefile` ?
all in all we should never use this mode for many days, main use for `-g` option is debugging or quick tests in test clusters or even `./start.sh -g | tee  rotatelogs`

2) looks not so bad. We need to make the change in [manage line 54](https://github.com/dmwm/deployment/blob/88e57d864473f7557884bdc8da947639c4e2f091/crabserver/manage#L54) somehow depend on whether we run with `-d` flag or not.

if we send output to stdout, do we still need rotatelogs ?

**Comment by belforte** [comment]
I am not sure how things will work in 2) after we restart crabserver with [/data/start.sh](https://github.com/dmwm/CMSKubernetes/blob/e3de55fa56b215dba7deb5d1d49c22a00a6a6936/docker/crabserver/start.sh). IIUC we stop the process which was writing to the pipe and start a new one which will do the same. Is that true ? Does it work ? IIUC when process which writes terminates, reader also close. So who will write to `PID 1` stdout ?

**Comment by belforte** [comment]
hmm.. maybe another `cat /mypipe` is started by the infinite loop ? sort of a cheap daemon !

**Comment by belforte** [comment]
the more I think about this, the more I like solution 2). What do you say @novicecpp , shall we call discussion closed and go ahead with testing ?

**Comment by novicecpp** [comment]
Yes!!!

**Comment by novicecpp** [comment]
Sorry for my laziness last Friday. I should put diff file (instead of copy/paste style) to make it clearer where it change for solution 2).
example with `/data/run.sh` file
https://github.com/novicecpp/CRABServer/blob/logs_stdout/cicd/logs_stdout/run.sh.diff

**Comment by novicecpp** [comment]
Let me answer your questions:

> in 1), can we work around the problem with GH mode by doing ./start.sh -g | tee /data/srv/logs/crabserver/somefile ?

Sure, you can do that but you need to do it every time or add functionality to `./start.sh` script.

> If we send output to stdout, do we still need rotatelogs ?

No. Stdout/err logs of the container are handled by kuberentes (where the log files are, how logs is rotate).
Reminder that k8s (actually, the container runtime) store stdout/stderr from container's PID1 only.

> I am not sure how things will work in 2) after we restart crabserver with [/data/start.sh](https://github.com/dmwm/CMSKubernetes/blob/e3de55fa56b215dba7deb5d1d49c22a00a6a6936/docker/crabserver/start.sh). IIUC we stop the process which was writing to the pipe and start a new one which will do the same. Is that true ? Does it work ? IIUC when process which writes terminates, reader also close. So who will write to PID 1 stdout ?

Yes, and it works. `cat` will terminates after crabserver process exits. But, we put `cat` inside an infinite loop [here](https://github.com/novicecpp/CRABServer/blob/742f67f4743370293fc8946c81d2988d0a3911c6/cicd/logs_stdout/run.sh.diff#L19-L23) like you said in another comment.

**Comment by belforte** [comment]
Don't we still need to start crond ?

**Comment by novicecpp** [comment]
No, we don't. 
Last time we used it was when we restarted the process every 4 days because memory leak. https://github.com/dmwm/deployment/commit/4d61704934fd18a3193cb0260e59cc021ed164f2

**Comment by novicecpp** [comment]
One problem that occurs when we write log to stdout is we cannot use filebeat to read log files and send them to logstash (For this dashboard) because we do not have log files in container space anymore.
Previously, logs are written to files inside the container, and we added a filebeat sidecar to read logs from the same path as the log written by CRAB REST.

After some discussion with @mrceyhun, I want to propose a solution:

**Read the CRAB REST stdout logs directly from k8s worker node.**

- Deploy new filebeat as `DaemonSet` and mount `/var/log/pods` from worker node to container as [hostPath](https://kubernetes.io/docs/concepts/storage/volumes/#hostpath) to read other container's stdout logs.
- Config filebeat to read logs from `crabserver` container only 
- Change CRAB REST loggger to append `Type=cherrypylog` and `Type=crablog` to every log messages.
- Need a small change in [logstash.conf](https://github.com/dmwm/CMSKubernetes/blob/85af5d15d4b781dc7e9a9af71640b9b21c028640/kubernetes/cmsweb/monitoring/crab/logstash.conf) to get pod name from path of log files.

This solution is currently deployed on test12. You can view the configuration I use here; filebeat configuration: [configmap.yaml](https://gitlab.cern.ch/tseethon/logstdout/-/blob/c4b62a371f5c57f84a7281afe8361f52250f089e/filebeat-stdout/configmap.yaml), [daemonset.yaml](https://gitlab.cern.ch/tseethon/logstdout/-/blob/c4b62a371f5c57f84a7281afe8361f52250f089e/filebeat-stdout/daemonset.yaml), and logstash config [configmap.yaml](https://gitlab.cern.ch/tseethon/logstdout/-/blob/c4b62a371f5c57f84a7281afe8361f52250f089e/logstash-test/configmap.yaml#L12-14).

I will deploy on testbed by this Friday.

**Comment by novicecpp** [comment]
## Recap the proposed solution on streaming all logs to the container's stdout.

**First**, we need to change crabrest to write all logs to stdout instead of write to file.
To do this, we create a new [named pipe](https://man7.org/linux/man-pages/man7/pipe.7.html) called `crabserver-fifo` and make crabrest write stdout to it.
Then, use `cat` to print the content of `crabserver-fifo` to stdout, which is the same stdout as container (because it is called by `run.sh`)
The following files need to change:

- [run.sh](https://github.com/dmwm/CMSKubernetes/blob/7f7773fec8344c445be639b78789204e34f7ff4e/docker/crabserver/run.sh); patch [run.sh.diff](https://gitlab.cern.ch/tseethon/logstdout/-/blob/b91ec672cabf4734f646c3e647ac33da7414e567/CRABServer/cicd/logs_stdout/run.sh.diff)
    - Add `mkfifo` step and start the infinite loop of `cat crabserver-fifo` at the end of script.
- [manage](https://github.com/dmwm/deployment/blob/439eaf4eb59b305ebfd5e5b93cd8dfde5999a85c/crabserver/manage); patch [manage.diff](https://gitlab.cern.ch/tseethon/logstdout/-/blob/b91ec672cabf4734f646c3e647ac33da7414e567/CRABServer/cicd/logs_stdout/manage.diff)
    - Change `wmc-httpd` argument to write logs to named pipe instead of `rotatelogs`.
- [WMCore/REST/Main.py](https://github.com/dmwm/WMCore/blob/9d8829f2fd4a79c1c423b9736cd08947878ef786/src/python/WMCore/REST/Main.py); patch [Main.py.diff](https://gitlab.cern.ch/tseethon/logstdout/-/blob/b91ec672cabf4734f646c3e647ac33da7414e567/CRABServer/cicd/logs_stdout/Main.py.diff)
    - Small change to `open()` mode in [line 406](https://github.com/dmwm/WMCore/blob/9d8829f2fd4a79c1c423b9736cd08947878ef786/src/python/WMCore/REST/Main.py#L406) to `a+b`, unless it will get `ValueError: can't have unbuffered text I/O`.

The advantages of this way are: we can start/stop the app inside the container without the container getting restarted (it is helpful for development) and only introduce a few code changes.

The new process execution tree would look like this:

![process_execution_stdout](https://user-images.githubusercontent.com/2346664/206710903-ea43bcca-f3b9-468e-9b3f-0c220e4fe025.png)

**Second**, we need to change filebeat sidecar (the one that sends to logstash) to read logs from the container's stdout instead of log files inside crabrest container.
To do this, we deploy a new filebeat daemonset, mount `/var/log/pods` from the worker node to the container, and config the reading path to only read crabrest logs.

Please see k8s manifests and filebeat/logstash config in https://gitlab.cern.ch/tseethon/logstdout/-/tree/b91ec672cabf4734f646c3e647ac33da7414e567/crabserver-stdout/

The deployment diagram to make it easier to understand:

![logsoutput.drawio.png](https://user-images.githubusercontent.com/2346664/206711078-127b7cad-9969-4cc1-a496-8bf3da83d6e7.png)

I have deployed this in testbed cluster, and it works nicely. Metrics correctly show in [CRAB crabserver timber](https://monit-grafana.cern.ch/goto/hQr-N3FVk?orgId=11) dashboard.

**Comment by novicecpp** [comment]
I have tested the new filebeat and collected resource usage stats to see how much resource is used.

The service is a simple [logsoutput](https://gitlab.cern.ch/tseethon/logstdout/-/blob/84b6a949ecac50794216e60259ff0a03127752b5/main.go) binary that randomly selects one line from the log file and prints to stdout.

In the test environment, I deployed 8 `logsoutput` pods on a single worker node in a private k8s cluster. I set the log rate to 2 times the production workload (around 100 lines/seconds, `crabserver` only produces 47 lines/seconds).
Usually, only 20% of logs are collected and sent to logstash. But in this test, I sent 80% of logs to see how well filebeat handles large logs volumes (see [configmap.yaml](https://gitlab.cern.ch/tseethon/logstdout/-/blob/master/filebeat-filterlog/configmap.yaml)).

Here is the CPU and Memory metrics from Grafana dashboard.

Test:

![Screenshot from 2022-12-12 18-38-19](https://user-images.githubusercontent.com/2346664/207115136-6e507a4a-6e0f-4d6c-931e-cc70aeedb46a.png)
![Screenshot from 2022-12-12 18-38-35](https://user-images.githubusercontent.com/2346664/207115154-66beded8-6347-4a4b-8a4c-2c8255a93dcb.png)


Current Production sidecar:

![Screenshot from 2022-12-12 18-34-14](https://user-images.githubusercontent.com/2346664/207114685-7ce527ab-6365-440c-ab26-ed564142a373.png)


In summary, it takes around 15% CPU and 110 MiB memory, compared to the sidecar solution we have right now in production that uses <1% CPU and 40 MiB. The significant CPU usage difference is from the large logs volume in the test environment I explain above.
Also, this is 1 filebeat (daemonset) in a single machine versus 8 filebeat (sidecar) in multiple machines.

**Comment by novicecpp** [comment]
I have posted a proposal above in [CMSKUBERNETES-215](https://its.cern.ch/jira/projects/CMSKUBERNETES/issues/CMSKUBERNETES-215).
